### PR TITLE
feat(downsample): support export to multiple paths in Iceberg format

### PR DESCRIPTION
**Pull request by alextheimer**
**Pull Request checklist**

- [x] The commit(s) message(s) follows the contribution [guidelines](CONTRIBUTING.md) ?
- [x] Tests for the changes have been added (for bug fixes / features) ?
- [ ] Docs have been added / updated (for bug fixes / features) ?

Adds support for export from the downsampler job to multiple destinations; data is now exported in the Iceberg format.
This PR will remove all support for export of CSV-formatted data. 

## From https://github.com/filodb/FiloDB/pull/1718:
**Current behavior :** 
Spark Downsampler job currently exports data in CSV format.

**New behavior :**
Implements changes in Spark Downsampler Job to export data in Iceberg format and write to Iceberg Table in a specific schema.

Changes includes the following:

- Updates in `BatchExporter.scala` to be able to be to create RDD in new schema and convert RDD to DF to be able to write to Iceberg Table. 
- `filodb-defaults.conf` changes for export includes the following: 

1. Adds catalog, database conf Iceberg properties in filodb-defaults.conf.
2. Under groups, each group should also include, additional dynamic time series label based columns in addition to fixed columns in Iceberg Table.  
3. Each group, should also include table name, table location.
4. Each group, can also specify, if partitioning needs to be done by some additional label based columns, 

- To implement above, additional changes are implemented in `DownsamplerSettings.scala`, `BatchExporter.scala`, `DownsamplerMain.scala`. 
- Implements a new ExportTableConfig Class to translate all data like table path, table names, export rules, per group from config. 
- ExportTableConfig is also used to generate ExportSchema dynamically.
- Adds create table and create database Iceberg methods.
- In Iceberg, data partitioning will be directly handled by the create table sql statement.
- Rename destination-format key in filodb-defaults.conf with destination-path.

Additional changes to remove old CSV based export: 
- Remove old CSV based partition-by-template handling to generate data partition keys.
- Remove CSV specific properties like options, path-spec from filodb-defaults.conf. As mentioned above, for Iceberg, path spec for partitioning is not required. It will be handled by Iceberg create table statement.
- Remove old CSV based make-label-string methods/properties. In Iceberg table, labels will be stored in column labels of type map<string, string>, hence, string labels are not required.

### CODE REVIEW

File: spark-jobs/src/test/scala/filodb/downsampler/DownsamplerMainSpec.scala [hunk 1]
```diff
@@ -400,93 +448,99 @@ class DownsamplerMainSpec extends AnyFunSpec with Matchers with BeforeAndAfterAl
...
-      // ======= COMBINATIONS =======
-      """ 'foo\" ' "'\'"\ bar 'baz" """ -> """{"key":" 'foo\\\" ' \"'\\'\"\\ bar 'baz\" "}""",
-      """ 'foo\" ' ,, "'\'"\ bar 'baz" \, """ -> """{"key":" 'foo\\\" ' ,, \"'\\'\"\\ bar 'baz\" \\, "}""",
-      """ ["foo","ba'r:1234"] """ -> """{"key":" [\"foo\",\"ba'r:1234\"] "}""",
-      """ "foo,bar:1234" """ -> """{"key":" \"foo,bar:1234\" "}"""
-    )
-    for ((value, expected) <- inputOutputPairs) {
-      val map = Map("key" -> value)
-      BatchExporter.makeLabelString(map) shouldEqual expected
+  it("should give correct export schema") {
```
Comment (sandeep6189) [inline spark-jobs/src/test/scala/filodb/downsampler/DownsamplerMainSpec.scala:497]: lovely change . so much simpler

File: spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerSettings.scala [hunk 2]
```diff
@@ -80,13 +82,39 @@ class DownsamplerSettings(conf: Config = ConfigFactory.empty()) extends Serializ
...
+      val tablePath = group.as[String]("table-path")
+      val labelColumnMapping = group.as[Seq[String]]("label-column-mapping")
+        .sliding(2, 2).map(seq => (seq.head, seq.last)).toSeq
+      // Constructs dynamic exportSchema as per ExportTableConfig.
+      // final schema is a combination of columns defined in conf file plus some
+      // additional standardized columns
+      val tableSchema = {
+        // NOTE: ArrayBuffers are sometimes used instead of Seq's because otherwise
+        //   ArrayIndexOutOfBoundsExceptions occur when Spark exports a batch.
+        val fields = new mutable.ArrayBuffer[StructField](labelColumnMapping.length + 9)
```
Comment (sandeep6189) [inline spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerSettings.scala:100]: nit: add a comment on my we are adding `+9`. Might be better if we add this as a constant and add a description there

File: spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerSettings.scala [hunk 3]
```diff
@@ -80,13 +82,39 @@ class DownsamplerSettings(conf: Config = ConfigFactory.empty()) extends Serializ
...
-
   @transient lazy val exportDropLabels = downsamplerConfig.as[Seq[String]]("data-export.drop-labels")
 
   @transient lazy val exportKeyToRules = {
     val keyRulesPairs = downsamplerConfig.as[Seq[Config]]("data-export.groups").map { group =>
       val key = group.as[Seq[String]]("key")
+      val tableName = group.as[String]("table")
+      val tablePath = group.as[String]("table-path")
+      val labelColumnMapping = group.as[Seq[String]]("label-column-mapping")
+        .sliding(2, 2).map(seq => (seq.head, seq.last)).toSeq
```
Comment (sandeep6189) [inline spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerSettings.scala:93]: for. my understanding, what does sliding (2, 2 ) does?

File: spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerMain.scala [hunk 4]
```diff
@@ -143,40 +178,43 @@ class Downsampler(settings: DownsamplerSettings) extends Serializable {
...
+        // downsample the data as the first key is exported
+        val headTask = Seq(() =>
+          exportForKey(rddWithDs, exportKeyToRules.head._1, exportKeyToRules.head._2, batchExporter, spark))
+        // export all remaining keys without the downsample step
+        val tailTasks = exportKeyToRules.tail.map{spec => () =>
+          exportForKey(rdd, spec._1, spec._2, batchExporter, spark)}
+        headTask ++ tailTasks
+      }
+      // export/downsample RDDs in parallel
+      exportTasks.par.foreach(_.apply())
```
Comment (sandeep6189) [inline spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerMain.scala:214]: should we add a parallel config to control the degree of parallelism ? what is the default behavior ?

File: spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerMain.scala [hunk 5]
```diff
@@ -75,6 +76,40 @@ class Downsampler(settings: DownsamplerSettings) extends Serializable {
   lazy val exportLatency =
     Kamon.histogram("export-latency", MeasurementUnit.time.milliseconds).withoutTags()
 
+  /**
+   * Exports an RDD for a specific export key.
```
Comment (sandeep6189) [inline spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerMain.scala:80]: kudos on the func doc 👍

File: spark-jobs/src/main/scala/filodb/downsampler/chunk/BatchExporter.scala [hunk 6]
```diff
@@ -149,60 +117,95 @@ case class BatchExporter(downsamplerSettings: DownsamplerSettings, userStartTime
...
 
   /**
    * Converts an ExportData to a Spark Row.
-   * The result Row *must* match this.exportSchema.
+   * The result Row *must* match exportSchema.
    */
-  private def exportDataToRow(exportData: ExportRowData): Row = {
-    val dataSeq = new mutable.ArrayBuffer[Any](3 + downsamplerSettings.exportPathSpecPairs.size)
+  private def exportDataToRow(exportData: ExportRowData, exportTableConfig: ExportTableConfig): Row = {
+    val dataSeq = new mutable.ArrayBuffer[Any](exportTableConfig.tableSchema.fields.length)
```
[the remark this answers is not in the document]
Comment (nikitag55) [reply spark-jobs/src/main/scala/filodb/downsampler/chunk/BatchExporter.scala:142]: Yes, this is the array size for dataSeq. This array size is dependent on the number of standard + dynamic columns in tableSchema.

File: spark-jobs/src/main/scala/filodb/downsampler/chunk/BatchExporter.scala [hunk 7]
```diff
@@ -149,60 +117,95 @@ case class BatchExporter(downsamplerSettings: DownsamplerSettings, userStartTime
...
+    s"""
+       |CREATE DATABASE IF NOT EXISTS $database
+       |""".stripMargin
+
+  private def sqlCreateTable(catalog: String, database: String, exportTableConfig: ExportTableConfig) = {
+    // create dynamic col names with type to append to create table statement
+    val dynamicColNames = exportTableConfig.labelColumnMapping.map(pair => pair._2 + " string").mkString(", ")
+    val partitionColNames = exportTableConfig.partitionByCols.mkString(", ")
+    s"""
+       |CREATE TABLE IF NOT EXISTS $catalog.$database.${exportTableConfig.tableName} (
```
[the remark this answers is not in the document]
Comment (nikitag55) [reply spark-jobs/src/main/scala/filodb/downsampler/chunk/BatchExporter.scala:183]: This is not really a string constant. This string is dynamically created as per the input params of this function.

File: spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerSettings.scala (hunk 2 above)
Comment (nikitag55) [reply spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerSettings.scala:106]: Added comment in the latest commit.

File: spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerSettings.scala (hunk 3 above)
Comment (nikitag55) [reply spark-jobs/src/main/scala/filodb/downsampler/chunk/DownsamplerSettings.scala:97]: Added comments to explain this sliding and to creation of labelColumnMapping Seq[(a,b), (c,d)]

**Review by nikitag55 — DISMISSED** [review]

**Review by nikitag55 — APPROVED** [review]

**Pull request merged**